sympy/matrices/expressions/tests/test_matexpr.py

```diff
@@ -404,6 +404,67 @@ def test_issue_21195():
     assert A.diff(x) == B
 
 
+def test_issue_24859():
+    A = MatrixSymbol('A', 3, 4)
+    B = MatrixSymbol('B', 4, 3)
+    J = A*B
+    Jinv = Matrix(J).adjugate()
+    u = MatrixSymbol('u', 3, 4)
+    Jk = Jinv.subs(A, A + x*u)
+
+    assert Jk[0,0].diff(x).simplify() == (x*u[1, 0] + A[1, 0])*B[0, 1]*B[1, 2]*u[2, 1] + (x*u[1, 0] + \
+            A[1, 0])*B[0, 1]*B[2, 2]*u[2, 2] + (x*u[1, 0] + A[1, 0])*B[0, 1]*B[3, 2]*u[2, 3] - \
+            (x*u[1, 0] + A[1, 0])*B[0, 2]*B[1, 1]*u[2, 1] - (x*u[1, 0] + A[1, 0])*B[0, 2]*B[2, 1]*u[2, 2] - \
+            (x*u[1, 0] + A[1, 0])*B[0, 2]*B[3, 1]*u[2, 3] - (x*u[1, 1] + A[1, 1])*B[0, 1]*B[1, 2]*u[2, 0] + \
```

Comment:
A simpler test case would be better.

We don't need such a complicated example to demonstrate the bug:
```python
In [6]: A = MatrixSymbol('A', 3, 3)

In [7]: B = MatrixSymbol('B', 3, 3)

In [8]: x = symbols("x")

In [9]: M00 = MatrixElement(A + x*B, 0, 0)

In [10]: M00
Out[10]: (x⋅B + A)[0, 0]

In [11]: M00.diff(x)
Out[11]: 0

In [12]: M00.doit().diff(x)
Out[12]: B₀₀
```
We could also use something like the original example but just reduce the size of the matrices:
```python
In [13]:     A = MatrixSymbol('A', 2, 3)
    ...:     B = MatrixSymbol('B', 3, 2)
    ...:     J = A*B
    ...:     Jinv = Matrix(J).adjugate()
    ...:     u = MatrixSymbol('u', 2, 3)
    ...:     Jk = Jinv.subs(A, A + x*u)

In [14]: Jk[0,0]
Out[14]: (x⋅u + A)[1, 0]⋅B₀₁ + (x⋅u + A)[1, 1]⋅B₁₁ + (x⋅u + A)[1, 2]⋅B₂₁

In [15]: Jk[0,0].diff(x)
Out[15]: 0

In [16]: Jk[0,0].doit().diff(x)
Out[16]: B₀₁⋅u₁₀ + B₁₁⋅u₁₁ + B₂₁⋅u₁₂
```